blockproducer/chain.go

```diff
@@ -45,224 +49,238 @@ var (
+	for i, v := range branches {
+		if head == nil || v.head.count > head.head.count {
+			headIndex = i
+			head = v
+		}
 	}
 
-	// get accountAddress
-	pubKey, err := kms.GetLocalPublicKey()
-	if err != nil {
-		return nil, err
+	// Get accountAddress
+	if pubKey, err = kms.GetLocalPublicKey(); err != nil {
+		return
 	}
-	accountAddress, err = crypto.PubKeyHash(pubKey)
-	if err != nil {
-		return nil, err
+	if addr, err = crypto.PubKeyHash(pubKey); err != nil {
+		return
 	}
 
-	chain = &Chain{
-		db:            db,
-		ms:            newMetaState(),
-		bi:            newBlockIndex(),
-		rt:            newRuntime(cfg, accountAddress),
-		cl:            rpc.NewCaller(),
-		blocksFromRPC: make(chan *pt.Block),
-		pendingTxs:    make(chan pi.Transaction),
-		stopCh:        make(chan struct{}),
+	// Setup peer list
+	if locSvIndex, ok = cfg.Peers.Find(cfg.NodeID); !ok {
+		err = ErrLocalNodeNotFound
+		return
+	}
+	if t = cfg.ComfirmThreshold; t <= 0.0 {
+		t = float64(2) / 3.0
+	}
+	if m = math.Ceil(l*t + 1); m > l {
+		m = l
 	}
 
-	err = chain.db.View(func(tx *bolt.Tx) (err error) {
-		meta := tx.Bucket(metaBucket[:])
-		metaEnc := meta.Get(metaStateKey)
-		if metaEnc == nil {
-			return ErrMetaStateNotFound
-		}
-
-		state := &State{}
-		if err = utils.DecodeMsgPack(metaEnc, state); err != nil {
-			return
-		}
-		chain.rt.setHead(state)
-
-		var last *blockNode
-		var index int32
-		blocks := meta.Bucket(metaBlockIndexBucket)
-		nodes := make([]blockNode, blocks.Stats().KeyN)
-
-		if err = blocks.ForEach(func(k, v []byte) (err error) {
-			block := &pt.Block{}
-			if err = utils.DecodeMsgPack(v, block); err != nil {
-				log.WithError(err).Error("load block failed")
-				return err
-			}
-
-			log.Debugf("load chain block %s, parent block %s", block.BlockHash(), block.ParentHash())
-
-			parent := (*blockNode)(nil)
-
-			if last == nil {
-				// check genesis block
-			} else if block.ParentHash().IsEqual(&last.hash) {
-				if err = block.SignedHeader.Verify(); err != nil {
-					return err
-				}
-
-				parent = last
-			} else {
-				parent = chain.bi.lookupNode(block.ParentHash())
+	// create chain
+	c = &Chain{
+		ctx:    cld,
+		cancel: ccl,
+		wg:     &sync.WaitGroup{},
 
-				if parent == nil {
-					return ErrParentNotFound
-				}
-			}
+		server: cfg.Server,
+		cl:     rpc.NewCaller(),
 
-			nodes[index].initBlockNode(block, parent)
-			chain.bi.addBlock(&nodes[index])
-			last = &nodes[index]
-			index++
-			return err
-		}); err != nil {
-			return err
-		}
+		st: st,
+		bs: bus,
 
-		// Reload state
-		if err = chain.ms.reloadProcedure()(tx); err != nil {
-			return
-		}
+		pendingBlocks: make(chan *types.BPBlock),
+		pendingTxs:    make(chan pi.Transaction),
 
-		return
-	})
-	if err != nil {
-		return nil, err
+		address:     addr,
+		genesisTime: cfg.Genesis.SignedHeader.Timestamp,
+		period:      cfg.Period,
+		tick:        cfg.Tick,
+
+		peers:      cfg.Peers,
+		nodeID:     cfg.NodeID,
+		comfirms:   uint32(m),
```

Comment:
m is a float64, uint32(m) may cause some vuln problems.

Reply:
Spelling mistake: `comfirms`  -> `confirms`